File: app/src/main/java/sample/mail/ru/httpsample/HttpAsyncRequest.java
package sample.mail.ru.httpsample;

import android.os.AsyncTask;

import java.lang.ref.WeakReference;
import java.util.Arrays;


public class HttpAsyncRequest extends AsyncTask<String, Integer, String> {
    private WeakReference<MainActivity.RequestListener> mListener;
    protected int mErrorStringID;

    public HttpAsyncRequest(MainActivity.RequestListener listener) {
        mListener = new WeakReference<>(listener);
    }

    @Override
    protected String doInBackground(String... params) {
        if (params != null && params.length > 0) {
            HttpRequest request = new HttpRequest(params[0]);
            int status = request.makeRequest();

            if (status == HttpRequest.REQUEST_OK) {
                return StringUtils.join(Arrays.asList(request.getHeaders(), request.getContent()), "\n");
            }
            else if (status == HttpRequest.REQUEST_REDIRECT) {
                return doInBackground(request.getRedirectURL());
            }
            else {
                mErrorStringID = request.getErrorStringId();
            }
        }
        else {
            mErrorStringID = R.string.too_few_params;
        }
        return null;
    }

    @Override
    protected void onPostExecute(String s) {
        if (!isCancelled()) {
            MainActivity.RequestListener l = mListener.get();
            if (l != null) {
                if (s != null) {
                    l.onRequestResult(s);
                }
                else {
                    l.onRequestError(mErrorStringID);
                }
            }
        }
    }
}

File: app/src/main/java/sample/mail/ru/httpsample/HttpAsyncRequest.kt
package sample.mail.ru.httpsample

import android.os.AsyncTask

open class HttpAsyncRequest(private val listener: (Pair<String, Int>) -> Unit) : AsyncTask<String, Int, Pair<String, Int>>() {

    override fun doInBackground(vararg params: String): Pair<String, Int> {
        if (params.isNotEmpty()) {
            val request = HttpRequest(params[0])
            val status = request.makeRequest()

            return when (status) {
                HttpRequest.REQUEST_OK -> Pair("".join(listOf(request.headers, request.content), "\n"), 0)
                HttpRequest.REQUEST_REDIRECT -> doInBackground(request.redirectURL)
                else -> Pair("", request.errorStringId)
            }
        }
        return Pair("", R.string.too_few_params)
    }

    override fun onPostExecute(data : Pair<String, Int>) {
        if (!isCancelled) {
            listener(data)
        }
    }
}
